File: agent/src/runtime_secrets.py
"""Strict loading of runtime secrets from environment values or locked files."""

from __future__ import annotations

import os
from collections.abc import Mapping
from pathlib import Path
import re
import stat


_NAME_PATTERN = re.compile(r"^[A-Z][A-Z0-9_]*$")


class RuntimeSecretError(RuntimeError, ValueError):
    """Raised when a runtime secret is missing, ambiguous, or unsafe."""


class _RuntimeSecretFileMissing(RuntimeSecretError):
    pass


def _validate_value(
    name: str,
    value: str,
    *,
    required: bool,
    default: str | None,
    max_bytes: int,
) -> str | None:
    if len(value.encode("utf-8")) > max_bytes:
        raise RuntimeSecretError(f"{name} exceeds the size limit")
    if value.endswith("\n"):
        value = value[:-1]
    if "\n" in value or "\r" in value:
        raise RuntimeSecretError(f"{name} must be a single line")
    if any(ord(character) < 32 or ord(character) == 127 for character in value):
        raise RuntimeSecretError(f"{name} contains control characters")
    if not value:
        if required:
            raise RuntimeSecretError(f"{name} is required")
        return default
    return value


def _read_locked_file(name: str, path_value: str, max_bytes: int) -> str:
    path = Path(path_value)
    if not path.is_absolute():
        raise RuntimeSecretError(f"{name}_FILE must be absolute")
    try:
        metadata = path.lstat()
    except FileNotFoundError as error:
        raise _RuntimeSecretFileMissing(
            f"{name}_FILE is unavailable"
        ) from error
    except OSError as error:
        raise RuntimeSecretError(f"{name}_FILE is unavailable") from error
    if stat.S_ISLNK(metadata.st_mode):
        raise RuntimeSecretError(f"{name}_FILE must not be a symlink")

    flags = os.O_RDONLY
    flags |= getattr(os, "O_CLOEXEC", 0)
    flags |= getattr(os, "O_NOFOLLOW", 0)
    try:
        descriptor = os.open(path, flags)
    except FileNotFoundError as error:
        raise _RuntimeSecretFileMissing(
            f"{name}_FILE is unavailable"
        ) from error
    except OSError as error:
        raise RuntimeSecretError(f"{name}_FILE is unavailable") from error
    try:
        metadata = os.fstat(descriptor)
        if not stat.S_ISREG(metadata.st_mode):
            raise RuntimeSecretError(f"{name}_FILE must be a regular file")
        if stat.S_IMODE(metadata.st_mode) & 0o077:
            raise RuntimeSecretError(
                f"{name}_FILE permissions must allow only the owner"
            )
        if metadata.st_size > max_bytes:
            raise RuntimeSecretError(f"{name}_FILE exceeds the size limit")
        payload = os.read(descriptor, max_bytes + 1)
        if len(payload) > max_bytes:
            raise RuntimeSecretError(f"{name}_FILE exceeds the size limit")
    finally:
        os.close(descriptor)
    try:
        return payload.decode("utf-8", errors="strict")
    except UnicodeDecodeError as error:
        raise RuntimeSecretError(f"{name}_FILE must contain UTF-8") from error


def read_runtime_secret(
    name: str,
    *,
    environ: Mapping[str, str] | None = None,
    required: bool = False,
    default: str | None = None,
    max_bytes: int = 16_384,
) -> str | None:
    """Read one bounded secret while rejecting ambiguous configuration."""
    if not _NAME_PATTERN.fullmatch(name):
        raise RuntimeSecretError("secret name is invalid")
    if max_bytes < 1:
        raise RuntimeSecretError("max_bytes must be positive")

    values = os.environ if environ is None else environ
    inline_present = name in values
    file_name = f"{name}_FILE"
    file_present = file_name in values and bool(values[file_name])
    if inline_present and file_present:
        raise RuntimeSecretError(
            f"{name} and {file_name} are both configured"
        )

    if file_present:
        try:
            value = _read_locked_file(name, values[file_name], max_bytes)
        except _RuntimeSecretFileMissing:
            if required:
                raise
            return default
    elif inline_present:
        value = values[name]
    else:
        if required:
            raise RuntimeSecretError(f"{name} is required")
        return default

    return _validate_value(
        name,
        value,
        required=required,
        default=default,
        max_bytes=max_bytes,
    )

File: dashboard/runtime-secrets.mjs
import {
  closeSync,
  constants,
  fstatSync,
  lstatSync,
  openSync,
  readFileSync,
} from 'node:fs'
import { isAbsolute } from 'node:path'


const NAME_PATTERN = /^[A-Z][A-Z0-9_]*$/


export class RuntimeSecretError extends Error {
  constructor(message) {
    super(message)
    this.name = 'RuntimeSecretError'
  }
}


function validateValue(name, value, { required, defaultValue, maxBytes }) {
  if (Buffer.byteLength(value, 'utf8') > maxBytes) {
    throw new RuntimeSecretError(`${name} exceeds the size limit`)
  }
  if (value.endsWith('\n')) {
    value = value.slice(0, -1)
  }
  if (value.includes('\n') || value.includes('\r')) {
    throw new RuntimeSecretError(`${name} must be a single line`)
  }
  for (const character of value) {
    const codePoint = character.codePointAt(0)
    if (codePoint < 32 || codePoint === 127) {
      throw new RuntimeSecretError(`${name} contains control characters`)
    }
  }
  if (!value) {
    if (required) {
      throw new RuntimeSecretError(`${name} is required`)
    }
    return defaultValue
  }
  return value
}


function readLockedFile(name, path, maxBytes) {
  if (!isAbsolute(path)) {
    throw new RuntimeSecretError(`${name}_FILE must be absolute`)
  }
  let metadata
  try {
    metadata = lstatSync(path)
  } catch (error) {
    throw new RuntimeSecretError(
      `${name}_FILE is unavailable`,
      { cause: error },
    )
  }
  if (metadata.isSymbolicLink()) {
    throw new RuntimeSecretError(`${name}_FILE must not be a symlink`)
  }

  let descriptor
  try {
    descriptor = openSync(
      path,
      constants.O_RDONLY | (constants.O_NOFOLLOW ?? 0),
    )
    metadata = fstatSync(descriptor)
    if (!metadata.isFile()) {
      throw new RuntimeSecretError(
        `${name}_FILE must be a regular file`,
      )
    }
    if ((metadata.mode & 0o077) !== 0) {
      throw new RuntimeSecretError(
        `${name}_FILE permissions must allow only the owner`,
      )
    }
    if (metadata.size > maxBytes) {
      throw new RuntimeSecretError(
        `${name}_FILE exceeds the size limit`,
      )
    }
    const payload = readFileSync(descriptor)
    if (payload.length > maxBytes) {
      throw new RuntimeSecretError(
        `${name}_FILE exceeds the size limit`,
      )
    }
    try {
      return new TextDecoder('utf-8', { fatal: true }).decode(payload)
    } catch (error) {
      throw new RuntimeSecretError(
        `${name}_FILE must contain UTF-8`,
        { cause: error },
      )
    }
  } catch (error) {
    if (error instanceof RuntimeSecretError) {
      throw error
    }
    throw new RuntimeSecretError(
      `${name}_FILE is unavailable`,
      { cause: error },
    )
  } finally {
    if (descriptor !== undefined) {
      closeSync(descriptor)
    }
  }
}


export function readRuntimeSecret(
  environ,
  name,
  {
    required = false,
    defaultValue = undefined,
    maxBytes = 16_384,
  } = {},
) {
  if (!NAME_PATTERN.test(name)) {
    throw new RuntimeSecretError('secret name is invalid')
  }
  if (!Number.isInteger(maxBytes) || maxBytes < 1) {
    throw new RuntimeSecretError('maxBytes must be a positive integer')
  }

  const inlinePresent = Object.hasOwn(environ, name)
  const fileName = `${name}_FILE`
  const filePresent = (
    Object.hasOwn(environ, fileName) && Boolean(environ[fileName])
  )
  if (inlinePresent && filePresent) {
    throw new RuntimeSecretError(
      `${name} and ${fileName} are both configured`,
    )
  }

  let value
  if (filePresent) {
    value = readLockedFile(name, environ[fileName], maxBytes)
  } else if (inlinePresent) {
    value = environ[name]
  } else {
    if (required) {
      throw new RuntimeSecretError(`${name} is required`)
    }
    return defaultValue
  }
  return validateValue(
    name,
    value,
    { required, defaultValue, maxBytes },
  )
}
